File: packages/provide-uterm-go/detection/loader.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package detection

import (
	"encoding/json"
	"fmt"
	"os"
)

// RulesPath is a filesystem path to a rules.json file. It exists so
// LoadRuleset can distinguish a path (Python's pathlib.Path) from an inline
// JSON string, matching the reference load_ruleset overloads.
type RulesPath string

// RuleSetFromJSON parses a RuleSet from a JSON byte slice (model_validate).
func RuleSetFromJSON(data []byte) (*RuleSet, error) {
	var rs RuleSet
	if err := json.Unmarshal(data, &rs); err != nil {
		return nil, err
	}
	return &rs, nil
}

// RuleSetFromJSONFile loads a RuleSet from a file. Mirrors
// RuleSet.from_json_file, including its error prefix.
func RuleSetFromJSONFile(path string) (*RuleSet, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		//nolint:staticcheck // ST1005: message kept byte-identical to the Python implementation
		return nil, fmt.Errorf("Failed to load rules from %s: %w", path, err)
	}
	rs, err := RuleSetFromJSON(data)
	if err != nil {
		//nolint:staticcheck // ST1005: message kept byte-identical to the Python implementation
		return nil, fmt.Errorf("Failed to load rules from %s: %w", path, err)
	}
	return rs, nil
}

// LoadRuleset loads a RuleSet from a *RuleSet (passthrough), a RulesPath
// (file), or a string (inline JSON). Faithful to the Python load_ruleset,
// including its error messages.
func LoadRuleset(source any) (*RuleSet, error) {
	switch src := source.(type) {
	case *RuleSet:
		return src, nil
	case RuleSet:
		return &src, nil
	case RulesPath:
		if _, err := os.Stat(string(src)); err != nil {
			//nolint:staticcheck // ST1005: message kept byte-identical to the Python implementation
			return nil, fmt.Errorf("Rules file not found: %s", string(src))
		}
		return RuleSetFromJSONFile(string(src))
	case string:
		rs, err := RuleSetFromJSON([]byte(src))
		if err != nil {
			//nolint:staticcheck // ST1005: message kept byte-identical to the Python implementation
			return nil, fmt.Errorf("Failed to parse rules: %w", err)
		}
		return rs, nil
	default:
		//nolint:staticcheck // ST1005: message kept byte-identical to the Python implementation
		return nil, fmt.Errorf("Failed to parse rules: unsupported source type %T", source)
	}
}

File: packages/provide-uterm-ts/src/detection/loader.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Loading a rule set from wherever it lives.
 *
 * Port of the Python module `provide.uterm.detection.loader`.
 */

import { readFileSync } from "node:fs";
import { parseRuleSet, type RuleSet, RuleValidationError } from "./rules.ts";

/** Options for {@link loadRuleSet}. */
export interface LoadRuleSetOptions {
  /** Read `source` as a path rather than as JSON text. */
  fromFile?: boolean;
}

/** Whether this is already a rule set rather than something to read. */
function isRuleSet(source: RuleSet | string): source is RuleSet {
  return typeof source !== "string";
}

/**
 * Read a rule set from JSON text, a file, or one already in hand.
 *
 * A missing file is named in the message, because an operator with several
 * rules files needs to know which one. Anything that fails to parse or fails
 * validation is reported the same way — the distinction between "not JSON"
 * and "not a rule set" is not one the caller can act on differently.
 *
 * @throws {RuleValidationError} On a missing file, unparseable text, or text
 *   that parses but is not a rule set.
 */
export function loadRuleSet(source: RuleSet | string, options: LoadRuleSetOptions = {}): RuleSet {
  if (isRuleSet(source)) {
    return source;
  }
  if (options.fromFile === true) {
    let text: string;
    try {
      text = readFileSync(source, "utf-8");
    } catch (error) {
      throw new RuleValidationError(`Rules file not found: ${source} (${(error as Error).message})`);
    }
    return loadRuleSet(text);
  }
  let data: unknown;
  try {
    data = JSON.parse(source);
  } catch (error) {
    // Named as a JSON fault rather than folded in with a validation one: text
    // that is not JSON and text that is valid JSON but not a rule set are
    // different mistakes, and an operator fixes them differently.
    throw new RuleValidationError(`Failed to parse rules: not JSON: ${(error as Error).message}`);
  }
  try {
    return parseRuleSet(data);
  } catch (error) {
    throw new RuleValidationError(`Failed to parse rules: ${(error as Error).message}`);
  }
}
